README.rst
Python Client for Cloud Tasks API
=================================

|stable| |pypi| |versions|

`Cloud Tasks API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-tasks.svg
   :target: https://pypi.org/project/google-cloud-tasks/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-tasks.svg
   :target: https://pypi.org/project/google-cloud-tasks/
.. _Cloud Tasks API: https://cloud.google.com/tasks/docs/
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/cloudtasks/latest
.. _Product Documentation:  https://cloud.google.com/tasks/docs/

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Cloud Tasks API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Cloud Tasks API.:  https://cloud.google.com/tasks/docs/
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-tasks


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-tasks

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Cloud Tasks API
   to see other available methods on the client.
-  Read the `Cloud Tasks API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Cloud Tasks API Product documentation:  https://cloud.google.com/tasks/docs/
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

[PyPI History][1]

[1]: https://pypi.org/project/google-cloud-tasks/#history

## [1.5.2](https://github.com/googleapis/python-tasks/compare/v1.5.1...v1.5.2) (2022-06-09)


### Bug Fixes

* **deps:** require protobuf<4.0.0 on v1 branch ([#268](https://github.com/googleapis/python-tasks/issues/268)) ([a2cc376](https://github.com/googleapis/python-tasks/commit/a2cc376830253bffcd73595b58697747de9c24be))

### [1.5.1](https://github.com/googleapis/python-tasks/compare/v1.5.0...v1.5.1) (2022-03-25)


### Bug Fixes

* **deps:** require google-api-core >= 1.31.5, >= 2.3.2 on v1 release ([#237](https://github.com/googleapis/python-tasks/issues/237)) ([507b44c](https://github.com/googleapis/python-tasks/commit/507b44cea6518496d1743cd5ca79652dafe7f972))

## [1.5.0](https://www.github.com/googleapis/python-tasks/compare/v1.4.0...v1.5.0) (2020-02-24)


### Features

* **tasks:** add support for stackdriver logging config; update retry config (via synth) ([#8](https://www.github.com/googleapis/python-tasks/issues/8)) ([70b597a](https://www.github.com/googleapis/python-tasks/commit/70b597a615c75976a4993ab223328d7cba3bd139))

## [1.4.0](https://www.github.com/googleapis/python-tasks/compare/v1.3.0...v1.4.0) (2020-02-06)


### Features

* **tasks:** undeprecate resource helper methods; add py2 deprecation warning; change default timeouts; add 3.8 tests; edit docstrings (via synth)([#10074](https://www.github.com/googleapis/python-tasks/issues/10074)) ([5577817](https://www.github.com/googleapis/python-tasks/commit/5577817fbe6435af03d862761fa08288b02cc69a))


### Bug Fixes

* **tasks:** change default timeout values; bump copyright year to 2020; change line breaks in docstrings (via synth) ([#10271](https://www.github.com/googleapis/python-tasks/issues/10271)) ([f68536d](https://www.github.com/googleapis/python-tasks/commit/f68536d95b6e320e4140d6720cc0c47c184dd694))
* **tasks:** deprecate resource name helper methods (via synth) ([#9864](https://www.github.com/googleapis/python-tasks/issues/9864)) ([ccf2cab](https://www.github.com/googleapis/python-tasks/commit/ccf2cabbe32d91988bd9456dc777622182beb658))

## 1.3.0

11-04-2019 10:06 PST

### Implementation Changes
- Add proto annotations (via synth) ([#9352](https://github.com/googleapis/google-cloud-python/pull/9352))

### New Features
- Add HTTP tasks, OAuth tokens, and OIDC tokens (via synth) ([#9588](https://github.com/googleapis/google-cloud-python/pull/9588))

### Documentation
- Tweak docstrings (via synth) ([#9433](https://github.com/googleapis/google-cloud-python/pull/9433))
- Disambiguate client requests from cloud task requests ([#9398](https://github.com/googleapis/google-cloud-python/pull/9398))
- Change requests intersphinx url (via synth) ([#9409](https://github.com/googleapis/google-cloud-python/pull/9409))
- Update documentation (via synth) ([#9069](https://github.com/googleapis/google-cloud-python/pull/9069))
- Remove compatibility badges from READMEs ([#9035](https://github.com/googleapis/google-cloud-python/pull/9035))
- Fix intersphinx reference to requests ([#9294](https://github.com/googleapis/google-cloud-python/pull/9294))
- Remove CI for gh-pages, use googleapis.dev for `api_core` refs. ([#9085](https://github.com/googleapis/google-cloud-python/pull/9085))

## 1.2.1

08-12-2019 13:50 PDT

### Implementation Changes
- Remove send/recv msg size limit (via synth). ([#8971](https://github.com/googleapis/google-cloud-python/pull/8971))

### Documentation
- Fix links to googleapis.dev ([#8998](https://github.com/googleapis/google-cloud-python/pull/8998))
- Update intersphinx mapping for requests. ([#8805](https://github.com/googleapis/google-cloud-python/pull/8805))

## 1.2.0

07-24-2019 17:41 PDT


### Implementation Changes
- Allow kwargs to be passed to create_channel (via synth). ([#8406](https://github.com/googleapis/google-cloud-python/pull/8406))

### New Features
- Add 'options_' argument to clients' 'get_iam_policy'; pin black version (via synth). ([#8660](https://github.com/googleapis/google-cloud-python/pull/8660))
- Add 'client_options' support, update list method docstrings (via synth). ([#8524](https://github.com/googleapis/google-cloud-python/pull/8524))

### Dependencies
- Bump minimum version for google-api-core to 1.14.0. ([#8709](https://github.com/googleapis/google-cloud-python/pull/8709))
- Update pin for 'grpc-google-iam-v1' to 0.12.3+. ([#8647](https://github.com/googleapis/google-cloud-python/pull/8647))

### Documentation

- Link to googleapis.dev documentation in READMEs. ([#8705](https://github.com/googleapis/google-cloud-python/pull/8705))
- Add compatibility check badges to READMEs. ([#8288](https://github.com/googleapis/google-cloud-python/pull/8288))
- Fix typo in README. ([#8606](https://github.com/googleapis/google-cloud-python/pull/8606))

### Internal / Testing Changes
- Add docs job to publish to googleapis.dev. ([#8464](https://github.com/googleapis/google-cloud-python/pull/8464))
- Declare encoding as utf-8 in pb2 files (via synth). ([#8366](https://github.com/googleapis/google-cloud-python/pull/8366))
- Add disclaimer to auto-generated template files (via synth).  ([#8330](https://github.com/googleapis/google-cloud-python/pull/8330))
- Suppress checking 'cov-fail-under' in nox default session (via synth).  ([#8254](https://github.com/googleapis/google-cloud-python/pull/8254))
- Fix coverage in 'types.py' (via synth). ([#8166](https://github.com/googleapis/google-cloud-python/pull/8166))
- Blacken noxfile.py, setup.py (via synth). ([#8134](https://github.com/googleapis/google-cloud-python/pull/8134))
- Add empty lines (via synth). ([#8074](https://github.com/googleapis/google-cloud-python/pull/8074))

## 1.1.0

05-14-2019 15:30 PDT

### Implementation Changes
- Remove log_sampling_ratio, add stackdriver_logging_config (via synth). ([#7950](https://github.com/googleapis/google-cloud-python/pull/7950))

### Documentation

noxfile.py
# -*- coding: utf-8 -*-
#
# Copyright 2018 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Generated by synthtool. DO NOT EDIT!

from __future__ import absolute_import
import os
import shutil

import nox


BLACK_VERSION = "black==19.3b0"
BLACK_PATHS = ["docs", "google", "tests", "noxfile.py", "setup.py"]

if os.path.exists("samples"):
    BLACK_PATHS.append("samples")

DEFAULT_PYTHON_VERSION = "3.8"

# Error if a python version is missing
nox.options.error_on_missing_interpreters = True


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION, "click<8.1")
    session.run("black", "--check", *BLACK_PATHS)
    session.run("flake8", "google", "tests")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black.

    Format code to uniform standard.
    """
    session.install(BLACK_VERSION, "click<8.1")
    session.run("black", *BLACK_PATHS)


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")


def default(session):
    # Install all test dependencies, then install this package in-place.
    session.install("mock", "pytest", "pytest-cov")
    session.install("-e", ".")

    # Run py.test against the unit tests.
    session.run(
        "py.test",
        "--quiet",
        "--cov=google.cloud.cloudtasks",
        "--cov=google.cloud",
        "--cov=tests.unit",
        "--cov-append",
        "--cov-config=.coveragerc",
        "--cov-report=",
        "--cov-fail-under=0",
        os.path.join("tests", "unit"),
        *session.posargs,
    )


@nox.session(python=["2.7", "3.5", "3.6", "3.7", "3.8"])
def unit(session):
    """Run the unit test suite."""
    default(session)


@nox.session(python=["2.7", "3.7"])
def system(session):
    """Run the system test suite."""
    system_test_path = os.path.join("tests", "system.py")
    system_test_folder_path = os.path.join("tests", "system")
    # Sanity check: Only run tests if the environment variable is set.
    if not os.environ.get("GOOGLE_APPLICATION_CREDENTIALS", ""):
        session.skip("Credentials must be set via environment variable")

    system_test_exists = os.path.exists(system_test_path)
    system_test_folder_exists = os.path.exists(system_test_folder_path)
    # Sanity check: only run tests if found.
    if not system_test_exists and not system_test_folder_exists:
        session.skip("System tests were not found")

    # Use pre-release gRPC for system tests.
    session.install("--pre", "grpcio")

    # Install all test dependencies, then install this package into the
    # virtualenv's dist-packages.
    session.install("mock", "pytest")

    session.install("-e", ".")

    # Run py.test against the system tests.
    if system_test_exists:
        session.run("py.test", "--quiet", system_test_path, *session.posargs)
    if system_test_folder_exists:
        session.run("py.test", "--quiet", system_test_folder_path, *session.posargs)


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.

    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=86")

    session.run("coverage", "erase")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx<3.0.0", "alabaster", "recommonmark", "Jinja2<3.1")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )
